pkg/controller/bastion/actuator_reconcile.go

```diff
@@ -0,0 +1,469 @@
+	patch := client.MergeFrom(bastion.DeepCopy())
+	bastion.Status.Ingress = endpoints.public
+	return a.client.Status().Patch(ctx, bastion, patch)
+}
+
+// IngressReady returns true if either an IP or a hostname or both are set.
+func IngressReady(ingress *corev1.LoadBalancerIngress) bool {
+	return ingress != nil && (ingress.Hostname != "" || ingress.IP != "")
+}
+
+// addressToIngress converts the IP address into a
+// corev1.LoadBalancerIngress resource. If both arguments are nil, then
+// nil is returned.
+func addressToIngress(dnsName *string, ipAddress *string) *corev1.LoadBalancerIngress {
+	var ingress *corev1.LoadBalancerIngress
+
+	if ipAddress != nil || dnsName != nil {
+		ingress = &corev1.LoadBalancerIngress{}
+		if dnsName != nil {
+			ingress.Hostname = *dnsName
+		}
+
+		if ipAddress != nil {
+			ingress.IP = *ipAddress
+		}
+	}
+
+	return ingress
+}
+
+func getInstanceEndpoints(a *actuator, opt *Options, ip string) (*bastionEndpoints, error) {
+	response, err := getInstance(a, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	if response == nil {
+		return nil, fmt.Errorf("compute instance can't be nil")
+	}
+
+	if len(response.Instances.Instance) == 0 || response.Instances.Instance[0].Status != "Running" {
+		return nil, fmt.Errorf("compute instance not ready yet")
+	}
+
+	endpoints := &bastionEndpoints{}
+	instance := response.Instances.Instance[0]
+	internalIP := instance.VpcAttributes.PrivateIpAddress.IpAddress[0]
+
+	if ingress := addressToIngress(nil, &internalIP); ingress != nil {
+		endpoints.private = ingress
+	}
+
+	if ingress := addressToIngress(nil, &ip); ingress != nil {
+		endpoints.public = ingress
+	}
+
+	return endpoints, nil
+}
+
+func ensureComputeInstance(a *actuator, logger logr.Logger, opt *Options, securityGroupID, imageID, instanceTypeID string) (string, error) {
+	response, err := getInstance(a, opt)
+	if err != nil {
+		return "", err
+	}
+
+	if len(response.Instances.Instance) > 0 && response.Instances.Instance[0].InstanceName == opt.BastionInstanceName {
+		return response.Instances.Instance[0].InstanceId, nil
+	}
+
+	logger.Info("creating new bastion compute instance")
+
+	instance, err := createInstance(a, opt, securityGroupID, imageID, instanceTypeID)
+	if err != nil {
+		return "", err
+	}
+
+	return instance.InstanceIdSets.InstanceIdSet[0], nil
+}
+
+func ensureSecurityGroup(a *actuator, opt *Options) (string, error) {
+	response, err := getSecurityGroup(a, opt, "")
+	if err != nil {
+		return "", err
+	}
+
+	if len(response.SecurityGroups.SecurityGroup) > 0 && response.SecurityGroups.SecurityGroup[0].SecurityGroupName == opt.SecurityGroupName {
+		logger.Info("Security Group found", "security group", opt.SecurityGroupName)
+		return response.SecurityGroups.SecurityGroup[0].SecurityGroupId, nil
+	}
+
+	logger.Info("creating Security Group")
+
+	createResponse, err := createSecurityGroup(a, opt)
+	if err != nil {
+		return "", err
+	}
+
+	if !createResponse.IsSuccess() {
+		return "", err
+	}
```

Comment:
thanks, I was using `IsSuccess` for the Instance check as well,  but it is more like a response is a success, which does not mean the instance is running or Ready. In here the security group is created instantly not like an instance that requires a long time to create.   remove this part to avoid misleading.  9376cc87